ArjunaCore/arjuna/classes/com/arjuna/ats/internal/arjuna/objectstore/slot/SlotStoreAdaptor.java

```diff
@@ -0,0 +1,349 @@
+import com.arjuna.ats.arjuna.objectstore.ObjectStoreAPI;
+import com.arjuna.ats.arjuna.objectstore.StateStatus;
+import com.arjuna.ats.arjuna.state.InputObjectState;
+import com.arjuna.ats.arjuna.state.OutputObjectState;
+import com.arjuna.ats.internal.arjuna.common.UidHelper;
+import com.arjuna.common.internal.util.propertyservice.BeanPopulator;
+
+import java.io.IOException;
+import java.io.SyncFailedException;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Adaptor class that wraps the SlotStore to make it look like an ObjectStore.
+ * Modelled on HornetqObjectStoreAdaptor.
+ *
+ * @author Jonathan Halliday (jonathan.halliday@redhat.com), 2020-03
+ */
+public class SlotStoreAdaptor implements ObjectStoreAPI {
+
+    private final SlotStore store;
+
+    // used for standalone bootstrap via StoreManager
+    public SlotStoreAdaptor() throws IOException {
+        SlotStoreEnvironmentBean envBean = BeanPopulator.getDefaultInstance(SlotStoreEnvironmentBean.class);
+        store = new SlotStore(envBean);
+    }
+
+    // used for beans wiring type bootstrap when running embedded.
+    public SlotStoreAdaptor(SlotStore slotStore) {
+        this.store = slotStore;
+    }
+
+    /**
+     * Obtain all of the Uids for a specified type.
+     *
+     * @param typeName       The type to scan for.
+     * @param foundInstances The object state in which to store the Uids
+     * @param matchState     The file type to look for (e.g., committed, shadowed). [StateStatus]
+     * @return <code>true</code> if no errors occurred, <code>false</code>
+     * otherwise.
+     */
+    @Override
+    public boolean allObjUids(String typeName, InputObjectState foundInstances, int matchState) throws ObjectStoreException {
+        if (tsLogger.logger.isTraceEnabled()) {
+            tsLogger.logger.trace("SlotStore.allObjUids(" + typeName + ", " + matchState + ")");
+        }
+
+        SlotStoreKey key = new SlotStoreKey(null, typeName, matchState);
+
+        SlotStoreKey[] matchingKeys = store.getMatchingKeys(key);
+
+        OutputObjectState buffer = new OutputObjectState();
+
+        try {
+            if (matchingKeys != null) {
+                for (SlotStoreKey matchingKey : matchingKeys) {
+                    if (matchingKey != null) {
+                        UidHelper.packInto(matchingKey.getUid(), buffer);
+                    }
+                }
+            }
+            UidHelper.packInto(Uid.nullUid(), buffer);
+        } catch (IOException e) {
+            throw new ObjectStoreException(e);
+        }
+
+        foundInstances.setBuffer(buffer.buffer());
+
+        return true;
+    }
+
+    @Override
+    public boolean allObjUids(String typeName, InputObjectState foundInstances) throws ObjectStoreException {
+        if (tsLogger.logger.isTraceEnabled()) {
+            tsLogger.logger.trace("SlotStore.allObjUids(" + typeName + ")");
+        }
+
+        return allObjUids(typeName, foundInstances, StateStatus.OS_UNKNOWN);
```

Comment:
I'd have expected this method to return all Uids regardless of their StateStatus value. But since it only returns `OS_UNKNOWN` ones will you include some javadoc for the method signature.

Reply:
Arg. That's actually a bug in getMatchingKeys - if should indeed return all by using
slotStoreKey.getStateStatus() == StateStatus.OS_UNKNOWN
rather than
key.getStateStatus() == StateStatus.OS_UNKNOWN
I should probably call them lookFor and candidateMatch or something else more distinctive.

Reply:
@jhalliday I was also expecting a javadoc entry for this method since it only return only OS_UNKNOWN states, why not all states and whatever the answer is shouldn't the javadoc say why?

Reply:
It returns all states - OS_UNKNOWN functions as a wildcard in this case. I think that started out as a quirk of the FileStore and then spread, it's not documented behaviour of the RecoveryStore interface, but I'm happy to make it so. The javadoc in the adaptor is mostly copy/paste of the interface javadoc just to help me remember how things should work, which is why that method didn't get any.